codegen/collector/collector.go

```diff
@@ -131,3 +155,133 @@ func (c *collector) findImportRelativeToRoot(absoluteRoot, importedProtoFile str
 	return possibleImportPaths[0], nil
 
 }
+
+type ImportsFetcher func(file string) ([]string, error)
+
+type ImportsExtractor interface {
+	FetchImportsForFile(protoFile string, importsFetcher ImportsFetcher) ([]string, error)
+}
+
+// thread-safe
+// The synchronizedImportsExtractor provides synchronized access to imports for a given proto file.
+// It provides 2 useful features for tree traversal:
+//	1. Imports for each file are cached, ensuring that if we attempt to access that file
+//		during traversal again, we do not need to duplicate work.
+//	2. If imports for a file are unknown, and simultaneous go routines attempt to load
+//		the imports, only 1 will execute and the other will block, waiting for the result.
+//		This reduces the number of times we open and parse files.
+type synchronizedImportsExtractor struct {
+	// cachedImports contains a map of fileImports, each indexed by their file name
+	cachedImports sync.Map
+
+	// protect access to activeRequests
+	activeRequestsMu sync.RWMutex
+	activeRequests   map[string]*importsFetchRequest
+}
+
+type fileImports struct {
+	imports []string
+	err     error
+}
+
+func NewSynchronizedImportsExtractor() ImportsExtractor {
+	return &synchronizedImportsExtractor{
+		activeRequests: map[string]*importsFetchRequest{},
+	}
+}
+
+func (i *synchronizedImportsExtractor) FetchImportsForFile(protoFile string, importsFetcher ImportsFetcher) ([]string, error) {
+	// Attempt to load the imports from the cache
+	cachedImports, ok := i.cachedImports.Load(protoFile)
+	if ok {
+		typedCachedImports := cachedImports.(*fileImports)
+		return typedCachedImports.imports, typedCachedImports.err
+	}
+
+	i.activeRequestsMu.Lock()
+
+	// Ensure that we do not wait forever
+	ctxWithTimeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
```

Comment:
nit: since this called with defer it will only `cancel()` after the select block below, which could potentially block forever

Reply:
we shouldn't need to invoke `cancel()` if we're just relying on the timeout